[SPARK-6577] [MLlib] [PySpark] SparseMatrix should be supported in PySpark

MechCoder opened the pull request:
Supporting of SparseMatrix in PySpark.

MechCoder committed:
[SPARK-6577] SparseMatrix should be supported in PySpark

MechCoder commented:
ping @mengxr ?

SparkQA commented:
[Test build #29695 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29695/consoleFull) for   PR 5355 at commit [`b6384fe`](https://github.com/apache/spark/commit/b6384fea2d5173fc50e986a7277ef07191532178).

SparkQA commented:
[Test build #29695 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29695/consoleFull) for   PR 5355 at commit [`b6384fe`](https://github.com/apache/spark/commit/b6384fea2d5173fc50e986a7277ef07191532178).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparseMatrix(object):`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29695/
Test PASSed.

MechCoder committed:
Renamed indices to rowIndices and indptr to colPtrs

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29704/
Test PASSed.

MechCoder committed:
Add support for CSR matrix

SparkQA commented:
[Test build #29744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29744/consoleFull) for   PR 5355 at commit [`db76caf`](https://github.com/apache/spark/commit/db76cafdf63ca8d1238fd793d8867ae67d5a0d60).

MechCoder commented:
@mengxr I have added support for the transpose flag also.

Also, I'm not sure where the support for scipy sparse matrices should be done? For example, _convert_to_vector is used when a scipy sparse matrix is used in place of a SparseMatrix (https://github.com/apache/spark/blob/master/python/pyspark/mllib/linalg.py#L278) 

I'm not sure it is needed in the current PR. Could you have a look?

SparkQA commented:
[Test build #29744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29744/consoleFull) for   PR 5355 at commit [`db76caf`](https://github.com/apache/spark/commit/db76cafdf63ca8d1238fd793d8867ae67d5a0d60).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparseMatrix(object):`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29744/
Test PASSed.

MechCoder commented:
@mengxr @davies Can some initial comments be given on this?

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -670,6 +668,23 @@ def toArray(self):
 
+    def toSparse(self):
+        """Convert to Sparse Format"""
+        numRows = self.numRows
```
This could be simplified:

~~~python
indices = np.nonzero(self.values)[0]
colPtrs = np.cumsum(np.bincount(indices / self.numRows + 1))
values = self.values[indices]
rowIndices = indices % self.numRows
return SparseMatrix(self.numRows, self.numCols, colPtrs, indices, values)
~~~

MechCoder replied:
This will fail if the last (few) columns are only zeros. I have made a minor change to this. Thanks!

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
                 all(self.values == other.values))
 
 
+class SparseMatrix(object):
```
extends `Matrix`

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+    """Sparse Matrix stored in CSC format."""
+    def __init__(self, numRows, numCols, colPtrs, rowIndices, values,
+                 isTransposed=False):
+        self.numRows = numRows
```
`Matrix.__init__(...)`

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+        self.numRows = numRows
+        self.numCols = numCols
+        self.isTransposed = isTransposed
+        if isinstance(colPtrs, basestring):
```
Shall we define a utility method for this conversion?

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+        self.numCols = numCols
+        self.isTransposed = isTransposed
+        if isinstance(colPtrs, basestring):
+            self.colPtrs = np.frombuffer(colPtrs, dtype=np.uint64)
```
`np.uint64` -> `np.int32` (to be consistent with Scala implementation)

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+        else:
+            self.colPtrs = np.asarray(colPtrs, dtype=np.uint64)
+
+        if self.isTransposed and self.colPtrs.size != numRows + 1:
```
It is cleaner to write

~~~python
if self.isTranspose:
  ...
else:
  ...
~~~

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+            self.values = np.frombuffer(values, dtype=np.float64)
+        else:
+            self.values = np.asarray(values, dtype=np.float64)
+
```
check the lengths of rowIndices and values are equal.

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+            self.numRows, self.numCols, self.colPtrs.tostring(),
+            self.rowIndices.tostring(), self.values.tostring(),
+            self.isTransposed
+        )
```
put `) after `self.isTransposed`

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+        if self.isTransposed:
+            j, i = i, j
+
+        nz = self.rowIndices[self.colPtrs[j]: self.colPtrs[j + 1]]
```
This block could be optimized (no temp array):

~~~python
colStart = self.colPtrs[j]
colEnd = self.colPtrs[j + 1]
ind = bisect_left(self.rowIndices, i, colStart, colEnd)
if ind < colEnd and self.rowIndices[ind] == i:
  return self.values[ind]
else:
  return 0.0
~~~

MechCoder replied:
Are you sure this is much more beneficial? I think the present code just creates a memory slice and not an array, and I suspect np.searchsorted might be faster than bisect_left. If you still think this should be changed, I shall change it.

mengxr replied:
I didn't test the performance. Usually it takes less than 10 lookups for sparse matrices. I just feel that this is slightly easier to read. It is okay to keep the current implementation. See my comment below.

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+        for ptr in xrange(self.colPtrs.size - 1):
+            startptr = self.colPtrs[ptr]
+            endptr = self.colPtrs[ptr + 1]
+            sparsearr[offset + self.rowIndices[startptr: endptr]] = \
```
put a space before `:`

MechCoder replied:
There should not be a whitespace before `:` according to PEP8 :P

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +702,100 @@ def __eq__(self, other):
+        if self.isTransposed:
+            order = 'C'
+        else:
+            order = 'F'
```
I think we should always use Fortran order. We can modify `_densify_values()` such that the output is always column-majored.

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -695,6 +804,13 @@ def dense(numRows, numCols, values):
         return DenseMatrix(numRows, numCols, values)
 
+    @staticmethod
+    def sparse(numRows, numCols, indptr, indices, values):
```
`indptr` -> `colPtrs`, `indices` -> `rowIndices` (must match Scala arg names)

mengxr commented on python/pyspark/mllib/tests.py:
```diff
@@ -143,6 +143,57 @@ def test_matrix_indexing(self):
+        expected[1, 0] = 1.0
+        expected[2, 0] = 2.0
+        expected[1, 2] = 4.0
+        expected[2, 2] = 5.0
```
We can create `expected` directly (maybe less code to write).

mengxr commented on python/pyspark/mllib/tests.py:
```diff
@@ -143,6 +143,57 @@ def test_matrix_indexing(self):
+        expected[0, 1] = 2.0
+        expected[1, 2] = 4.0
+        expected[2, 0] = 9.0
+        expected[2, 2] = 8.0
```
ditto

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -670,6 +668,23 @@ def toArray(self):
         return self.values.reshape((self.numRows, self.numCols), order='F')
 
+    def toSparse(self):
+        """Convert to Sparse Format"""
```
`to sparse matrix.`

MechCoder committed:
Made the following changes
1. Used convert_to_array for array conversion.
2. Used F order for toArray
3. Minor improvements in speed.

MechCoder commented:
@mengxr I have addressed all your comments, except the one-inline one.

SparkQA commented:
[Test build #29939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29939/consoleFull) for   PR 5355 at commit [`3d2700e`](https://github.com/apache/spark/commit/3d2700e95d09a52956648c956747a7e80c33f4dc).

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +703,87 @@ def __eq__(self, other):
+            return 0.0
+        ind = np.searchsorted(nz, i)
+        if i == nz[ind]:
+            return self.values[self.colPtrs[j]: self.colPtrs[j + 1]][ind]
```
`nz[ind]` may go out of bound. I would do this:

~~~python
ind = np.searchsorted(nz, i) + colStart
if ind < colEnd and self.indices[ind] == i:
  return self.values[ind]
else:
  return 0.0
~~~

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +703,87 @@ def __eq__(self, other):
+
+        nz = self.rowIndices[self.colPtrs[j]: self.colPtrs[j + 1]]
+        if nz.size == 0 or i > nz[-1]:
+            return 0.0
```
This `if` is not necessary. `np.searchsorted` would implement the same logic (with little overhead).

MechCoder committed:
Check bounds for indexing

MechCoder commented:
@mengxr Thanks! I've fixed up your comment.

SparkQA commented:
[Test build #29948 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29948/consoleFull) for   PR 5355 at commit [`61a864a`](https://github.com/apache/spark/commit/61a864a7f621f7d94132e891e89734acdb8e285b).

SparkQA commented:
[Test build #29947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29947/consoleFull) for   PR 5355 at commit [`454ef2c`](https://github.com/apache/spark/commit/454ef2ca83ccd0046a8a1e714b86505f8059fda9).

SparkQA commented:
[Test build #29951 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29951/consoleFull) for   PR 5355 at commit [`ea2c54b`](https://github.com/apache/spark/commit/ea2c54bd7f1639b596f200157732d2888bf21e4a).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29946/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29950/
Test FAILed.

SparkQA commented:
[Test build #29948 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29948/consoleFull) for   PR 5355 at commit [`61a864a`](https://github.com/apache/spark/commit/61a864a7f621f7d94132e891e89734acdb8e285b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29948/
Test FAILed.

SparkQA commented:
[Test build #29947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29947/consoleFull) for   PR 5355 at commit [`454ef2c`](https://github.com/apache/spark/commit/454ef2ca83ccd0046a8a1e714b86505f8059fda9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparseMatrix(Matrix):`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29947/
Test FAILed.

SparkQA commented:
[Test build #29951 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29951/consoleFull) for   PR 5355 at commit [`ea2c54b`](https://github.com/apache/spark/commit/ea2c54bd7f1639b596f200157732d2888bf21e4a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparseMatrix(Matrix):`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29951/
Test FAILed.

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -670,6 +674,18 @@ def toArray(self):
+    def toSparse(self):
+        """Convert to SparseMatrix"""
+        indices = np.nonzero(self.values)[0]
+        colCount = np.bincount(indices / self.numRows + 1)
```
On second thought, this might be more readable:

~~~python
colCounts = np.bincount(indices / self.numRows)
colPtrs = np.cumsum(np.hstack((0, colCounts, np.zeros(self.numCols - colCounts.size))))
~~~

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +703,87 @@ def __eq__(self, other):
+        if i < 0 or i >= self.numRows:
+            raise ValueError("Row index %d is out of range [0, %d)"
+                             % (i, self.numRows))
+        if j >= self.numCols or j < 0:
```
`if j < 0 or j >= self.numCols` (to be consistent)

mengxr commented on python/pyspark/mllib/linalg.py:
```diff
@@ -687,6 +703,87 @@ def __eq__(self, other):
+        sparsearr = np.zeros(self.numRows * self.numCols, dtype=np.float64)
+
+        for ptr in xrange(self.colPtrs.size - 1):
+            startptr = self.colPtrs[ptr]
```
This part can be simplified too.

~~~
A = np.zeros((self.numRows, self.numCols), dtype=np.float64, order='F')
for k in xrange(self.colPtrs.size - 1):
    startptr = self.colPtrs[k]
    endptr = self.colPtrs[k + 1]
    if self.isTransposed:
        A[k, self.rowIndices[startptr:endptr]] == self.values[startptr:endptr]
   else:
        A[self.rowIndices[startptr:endptr], k] == self.values[startptr:endptr]
   return A.reshape(self.numRows * self.numCols, order='F')
~~~

MechCoder replied:
Thanks. I had a similar idea in mind, but I wrote this to prevent code duplication across `toArray` and `toDense`. In the above code, I'll have to reshape twice on converting to dense. Is that ok?

MechCoder replied:
Never mind. I have a better idea. Will fix it up now.

MechCoder committed:
More readable code for densifying

MechCoder commented:
@mengxr fixed! anything else?

SparkQA commented:
[Test build #30005 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30005/consoleFull) for   PR 5355 at commit [`7492190`](https://github.com/apache/spark/commit/749219013731436ac64b7578d2e3afef83dc154f).

mengxr commented:
LGTM pending Jenkins.

SparkQA commented:
[Test build #30005 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30005/consoleFull) for   PR 5355 at commit [`7492190`](https://github.com/apache/spark/commit/749219013731436ac64b7578d2e3afef83dc154f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparseMatrix(Matrix):`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30005/
Test PASSed.

mengxr commented:
Merged into master. Thanks!

MechCoder commented:
@mengxr I'm really disappointed that I could not get the code right in the first attempt. Hopefully I shall improve in future PR's. Thanks!
